tests/update/patch_before_migration.pm

```diff
@@ -73,6 +79,53 @@ sub patching_sle {
     set_var("SYSTEM_PATCHED", 1);
 }
 
+sub install_package {
+    my @pk_list = split(/,/, get_var('PACKAGES'));
+    for my $pk (@pk_list) {
+        zypper_call "in -t package $pk";
+    }
+}
+
+sub install_patterns {
+
+    my $pcm = 0;
+    my @pt_list;
+    my @pt_list_un;
+    my @pt_list_in;
+
+    # install all patterns from product.
+    if (check_var('PATTERNS', 'all')) {
+        @pt_list_un = split(/ /, script_output("zypper pt -u | grep '^ ' | awk '{print \$2}' | sort -u | xargs"));
+        @pt_list_in = split(/ /, script_output("zypper pt -i | grep '^i' | awk -F '|' '{print \$2}' | sort -u | xargs"));
+        my %installed_pt = ();
+        foreach (@pt_list_in) {
+            $installed_pt{$_} = 1;
+        }
+        @pt_list = sort grep(!$installed_pt{$_}, @pt_list_un);
+        $pcm = grep /Amazon-Web-Services|Google-Cloud-Platform|Microsoft-Azure/, @pt_list_in;
+    }
+    #install certain pattern from parameter.
+    else {
+        @pt_list = split(/,/, get_var('PATTERNS'));
+    }
+
```

Comment:
There will be a failure if Amazon-Web-Services is already installed on hdd image and I define PATTERNS=Microsoft-Azure here. Because you check whether pattern is installed only when PATTERNS=all
Seems you need move the code (for checking installed patterns) out of condition "if (check_var('PATTERNS', 'all'))"?

Reply:
I searched existing code, seems PATTERNS could be "default", such as:
products/sle/main.pm:            set_var('PATTERNS',   'default,pcm');

As to "pcm", I have no idea, maybe just let it fails if defined in migration test suite.

Reply:
I suppose the better soultion for this circumstances is uninstall certain pattern before intsall new one. 
Try to install Microsoft-Azure on a system with Amazon-Web-Services will always failed with no doubt.